src/tdastro/base_models.py

```diff
@@ -32,20 +53,85 @@ def __init__(self, host=None, ra=None, dec=None, distance=None, **kwargs):
+        self.add_parameter("distance", distance)
+
+    def add_parameter(self, name, value=None, required=False, **kwargs):
+        """Add a single parameter to the PhysicalModel. Checks multiple sources
+        in the following order:
+            1. Manually specified ``value``
+            2. An entry in ``kwargs``
+            3. An entry in the object's host.
+            4. ``None``
+        Sets an initial value for the attribute based on the given information.
+
+        Parameters
+        ----------
+        name : `str`
+            The parameter name to add.
+        value : any, optional
+            The information to use to set the parameter. Can be a constant
+            or a function.
+        required : `bool`
+            Fail if the parameter is set to ``None``.
+        **kwargs : `dict`, optional
+           All other keyword arguments, possibly including the parameter setters.
 
-        # Set RA, dec, and distance from the given value or, if it is None and
-        # we have a host, from the host's value.
-        self.ra = ra
-        self.dec = dec
-        self.distance = distance
-        if ra is None and host is not None:
-            self.ra = host.ra
-        if dec is None and host is not None:
-            self.dec = host.dec
-        if distance is None and host is not None:
-            self.distance = host.distance
+        Raises
+        ------
+        Raise a ``KeyError`` if there is a parameter collision.
+        Raise a ``ValueError`` if the parameter is required, but set to None.
+        """
+        if hasattr(self, name):
+            raise KeyError(f"Duplicate parameter set: {KeyError}")
+
+        if value is None and name in kwargs:
+            # The value wasn't set, but the name is in kwargs.
+            value = kwargs[name]
+        if value is not None:
+            if isinstance(value, types.FunctionType):
+                self.setters[name] = (ParameterSource.FUNCTION, value)
+                setattr(self, name, value(**kwargs))
+            else:
+                self.setters[name] = (ParameterSource.CONSTANT, value)
+                setattr(self, name, value)
+        elif self.host is not None and hasattr(self.host, name):
+            self.setters[name] = (ParameterSource.HOST, name)
+            setattr(self, name, getattr(self.host, name))
+        elif not required:
+            self.setters[name] = (ParameterSource.CONSTANT, None)
+            setattr(self, name, None)
+        else:
+            raise ValueError(f"Missing required parameter {name}")
+
+    def sample_parameters(self, **kwargs):
+        """Sample the model's underlying parameters if they are provided by a function.
 
-        self.effects = []
+        Parameters
+        ----------
+        **kwargs : `dict`, optional
+           All the keyword arguments, including the values needed to sample parameters.
+        """
+        if self.host is not None:
+            self.host.sample_parameters(**kwargs)
```

Comment:
I'm a little confused why we need a separate recipe for host and can't just do 
```
host = PhysicalModel()
host_parameters = host.sample_parameters()
```

Reply:
My thinking is that when we resampled the `PhysicalModel`'s parameters, we would always want to make sure to resampled the host's parameters as well. However if we'd prefer to separate the two, then we could use the approach you suggest. What would be the most useful from the user perspective?

Reply:
We can definitely change the name. The basic idea behind special casing host is to allow: 1) the source to read information that isn't provided in the source (RA, dec, etc.) and 2) provide a name for sources to refer back to their host. We could rename it to something like "background" or get rid of the special casing altogether and just have it be an attribute.